## IIP314W Optimización aplicada a negocios

Metodo Simplex - Matricial

### Problema de Optimización General

$$ Min \quad c \cdot x $$

$$s.a, \quad A x \le b$$

$$ x \ge 0$$

Ya vimos la forma de resolución de este problema mediante el algoritmo simplex tableau, ahora realizaremos lo mismo mediante el uso de operaciones matriciales. Al igual que en el caso anterior, partimos ingresando las variables de holgura, entonces al inicio tenemos lo siguiente:

| V.Básica | Ecuación | $\vec{x}$ | $\vec{x_B}$ | LD |
|:-:|:-:|:-:|:-:|:-:|
| z | F.Obj. | $c$ | 0 | 0 |
| $\vec{x_B}$ | Rest. | $A$ | $I$ | $b$ |

Donde $x_B$ son las variables de holgura creadas en el inicio. Veamos si es óptimo.

#### Prueba de optimalidad

$$ c_B \cdot B^{-1} A - c $$

si algún coeficiente es negativo (para máximo, solo cambia el signo para minimo), la solución no es óptima. En nuestro caso, $c_B$ inicial es $\vec{0}$ y $B^{-1} = I$, por lo que el resultado es $c$.

En general,

| V.Básica | Ecuación | $\vec{x^*}$ | LD |
|:-:|:-:|:-:|:-:|
| z | F.Obj. | $c^*$ | 0 |
| $\vec{x_B}$ | Rest. | $A^*$ | $b$ |

Ahora "*" es nuestro estandar.

#### Iteración

Se elige el coeficiente más negativo y este entra a la base, entonces cambiemos la columna para la entrada de variable original a la base.

| V.Básica | Ecuación | $\vec{x}$ | LD |
|:-:|:-:|:-:|:-:|
| z | F.Obj. | $c_B B^{-1}A-c$ | $c_B B^{-1}b$ |
| $\vec{x_B}$ | Rest. | $B^{-1} A$ | $B^{-1}b$ |

Entonces, el problema es conseguir $B^{-1}$, aplicamos el método al mismo ejemplo.

### Ejercicio 17

Su empresa vende diariamente dos tipos de computadores, el computador a lo vende a 900 dolares y el b que es la versión barata a 500 dolares. En el armado del computador a utiliza 4 horas y en el computador b usa una hora, tiene un total de 10 horas al día para armar computadores. Usted gasta en componentes 400 dolares para el producto A y 200 dolares para el producto b y tiene para invertir diariamente 1200 dolares. Adicionalmente tiene un mostrador, donde puede poner a lo más 4 computadores, no tiene bodega por lo que es el limite de inventario. Suponga que todo se vende diariamente, formule y solucione el modelo de programación lineal que defina la producción diaria.

#### Algoritmo Simplex Matricial

El modelo del tableau (para max) del problema en este caso.

| Tableau | $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | R |
|---|---:|---:|---:|---:|---:|---:|
| CR | 500 | 300 | 0 |0 |0 | 0|
| $x_3$ | 4 | 2 | 1 |0 |0 | 12|
| $x_4$ | 4 | 1 | 0 |1 |0 | 10|
| $x_5$ | 1 | 1 | 0 |0 |1 | 4|

Entonces, la base inicial es $x_3$, $x_4$ y $x_5$, por lo que $B = I$, entonces $B^{-1} = I$.

#### Paso 1 algoritmo:

- Optimalidad: Calcular $c_B B^{-1}A-c$

- $c_B = [0 \quad 0 \quad 0]$
- $B^{-1} = I_3 $
- $A = \begin{bmatrix}
4 & 2 \\
4 & 1 \\
1 & 1
\end{bmatrix}$
- $c = [500 \quad 300]$

Entonces el resultado es:

$c_B B^{-1}A-c = [-500 \quad -300]$, no es óptimo, entra $x_1$ a la base.

$x_4$ sale de la base, del mismo modo que el algoritmo tableau.

#### Paso 2 algoritmo

Actualizamos los calculos. $Base = [x_3, x_1, x_5], B_A = [x_4, x_2]$

- $c_B = [0 \quad 500 \quad 0]$
- $B = \begin{bmatrix}
1 & 4 & 0 \\
0 & 4 & 0 \\
0 & 1 & 1
\end{bmatrix}$
- $A = \begin{bmatrix}
0 & 2 \\
1 & 1 \\
0 & 1
\end{bmatrix}$
- $c = [0 \quad 300]$


In [33]:
import numpy as np
from numpy.linalg import inv

B = np.array([[1, 4, 0], [0, 4, 0], [0, 1, 1]])
Binv = inv(B)
Binv

array([[ 1.  , -1.  ,  0.  ],
       [ 0.  ,  0.25,  0.  ],
       [ 0.  , -0.25,  1.  ]])

In [34]:
A = np.array([[0, 2], [1, 1], [0, 1]])
cb = np.array([[0, 500, 0]])
c = np.array([0, 300])
b = np.array([12, 10, 4])

cb@Binv@A-c

array([[ 125., -175.]])

Como -175 es negativo, no tenemos la condición de optimalidad,  por lo que $x_2$ entra a la base. Veamos quien sale, debemos calcular $B^{-1} A$ y $B^{-1} b$.

In [35]:
Binv@A

array([[-1.  ,  1.  ],
       [ 0.25,  0.25],
       [-0.25,  0.75]])

In [36]:
Binv@b

array([2. , 2.5, 1.5])

In [37]:
cb@Binv@b

array([1250.])

Se saca el minimo entre 2/1, 2.5/0.25, 1.5/0.75. Entonces es 2, 10, 2. Se elige cualquiera, por lo que $x_2$ entra a la base $x_5$ sale de ella.

#### Paso 3 algoritmo

Actualizamos los calculos. $Base = [x_3, x_1, x_2], B_A = [x_4, x_5]$

- $c_B = [0 \quad 500 \quad 300]$
- $B = \begin{bmatrix}
1 & 4 & 2 \\
0 & 4 & 1 \\
0 & 1 & 1
\end{bmatrix}$
- $A = \begin{bmatrix}
0 & 0 \\
1 & 0 \\
0 & 1
\end{bmatrix}$
- $c = [0 \quad 0]$


In [61]:
B = np.array([[1, 4, 2], [0, 4, 1], [0, 1, 1]])
Binv = inv(B)
Binv

array([[ 1.        , -0.66666667, -1.33333333],
       [ 0.        ,  0.33333333, -0.33333333],
       [ 0.        , -0.33333333,  1.33333333]])

In [67]:
A = np.array([[0, 0], [1, 0], [0, 1]])
cb = np.array([[0, 500, 300]])
c = np.array([0, 0])
b = np.array([12, 10, 4])

cb@Binv@A-c

array([[ 66.66666667, 233.33333333]])

Como no hay negativos, cumple optimalidad.

In [68]:
Binv@A

array([[-0.66666667, -1.33333333],
       [ 0.33333333, -0.33333333],
       [-0.33333333,  1.33333333]])

In [69]:
Binv@b

array([8.8817842e-16, 2.0000000e+00, 2.0000000e+00])

In [70]:
cb@Binv@b

array([1600.])

Entonces, la solución es (2,2) y el optimo es 1600.

### Ejercicio 15

Un avión de carga tiene tres compartimientos para almacenar la carga: frontal, central y trasero. Estos compartimientos tienen una capacidad limitada tanto en peso como en volumen de carga, de acuerdo a la siguiente tabla:

|Compartimiento |Peso máximo (ton) |Volumen máximo (m3)|
|--|:--:|:--:|
|Frontal |10 |6800|
|Central |16 |8700|
|Trasero |8 |5300|

Para mantener la estabilidad del avión, la carga transportada debe estar distribuída de manera que, respecto a los pesos transportados, se utilice la misma proporción de la capacidad de peso en los tres compartimientos. Para el próximo vuelo, están disponibles cuatro cargas con las características que se describen en lan tabla:

|Carga |Peso (ton.) |Volumen (m3/ton) |Beneficio (M$/ton)|
|-|:-:|:-:|:-:|
|C1 |18 |480 |310|
|C2 |15 |650 |380|
|C3 |23 |580 |350|
|C4 |12 |390 |285|

Cualquier proporción de estas cargas puede ser aceptada. Considerando la información provista, formule un modelo de programación lineal que permita a la
empresa decidir cuánto aceptar de cada carga de manera de maximizar el beneficio total obtenido. Resolver.

In [3]:
from scipy.optimize import minimize
import numpy as np

def fo(x):
    factores = np.array([18*310,18*310,18*310,15*380,15*380,15*380,23*350,23*350,23*350,12*285,12*285,12*285])
    return -factores@x

r1 = lambda x: -1*((x[0]*480+x[3]*650+x[6]*580+x[9]*390) - 6800)
r2 = lambda x: -1*((x[1]*480+x[4]*650+x[7]*580+x[10]*390) - 8700)
r3 = lambda x: -1*((x[2]*480+x[5]*650+x[8]*580+x[11]*390) - 5300)
r4 = lambda x: -1*((x[0]*18+x[3]*15+x[6]*23+x[9]*12) - 10)
r5 = lambda x: -1*((x[1]*18+x[4]*15+x[7]*23+x[10]*12) - 16)
r6 = lambda x: -1*((x[2]*18+x[5]*15+x[8]*23+x[11]*12) - 8)
r7 = lambda x: -1*(sum(x[0:3]) - 1)
r8 = lambda x: -1*(sum(x[3:6]) - 1)
r9 = lambda x: -1*(sum(x[6:9]) - 1)
r10 = lambda x: -1*(sum(x[9:12]) - 1)

e1 = lambda x: (sum(x[::3])/10 - sum(x[1::3])/16)
e2 = lambda x: (sum(x[::3])/10 - sum(x[2::3])/8)

# si hay mas restricciones se agregan a la lista
rest_list = [{'type': 'ineq', 'fun': r1}, {'type': 'ineq', 'fun': r2}, {'type': 'ineq', 'fun': r3}, \
             {'type': 'ineq', 'fun': r4}, {'type': 'ineq', 'fun': r5}, {'type': 'ineq', 'fun': r6}, \
             {'type': 'ineq', 'fun': r7}, {'type': 'ineq', 'fun': r8}, {'type': 'ineq', 'fun': r9}, \
             {'type': 'ineq', 'fun': r10}, {'type': 'eq', 'fun': e1}, {'type': 'eq', 'fun': e2}]

# las nunca bien ponderadas restricciones de no negatividad
noneg = [(0, 1)]*12

solucion_inicial = np.zeros(12)

# Minimize the function using the Nelder-Mead method
result = minimize(fo, solucion_inicial, method='trust-constr', bounds=noneg, constraints=rest_list)

# Print the result
print(result)


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.5.1 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Alumno\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "C:\Users\Alumno\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1082, in launch_instance
    app.start()
  File "C:\Users\Alumno\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 758, in start
    self.io_loop.start()
  File "C:\Users\Alumno\anaconda3\Lib\site-

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.5.1 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.



ImportError: numpy.core.multiarray failed to import

### Ejercicio 16

Una importadora distribuye un único producto en Chile. La empresa puede recibir el producto en **I** puertos en el país y desde los puertos envía directamente los productos a sus **J** clientes. El cliente $j$ tiene una demanda de $B_j$ unidades. La empresa enfrenta costos fijos de valor $F_i$ por utilizar el puerto $i$ para recibir productos y costos unitarios de transporte $C_{ij}$ por unidad enviada del puerto $i$ al cliente $j$. En caso de ser
utilizado, el puerto $i$ tiene capacidad para recibir y enviar hasta $A_i$ unidades del producto.

- Formule un modelo de programación lineal entera mixta que permita a la empresa determinar qué puertos utilizar para importar los productos y enviar a sus clientes para satisfacer la demanda de manera de minimizar los costos totales, es decir la suma de los costos fijos de usar los puertos y los costos de transporte.

- Considere ahora que no se puede enviar al mismo cliente productos desde más de dos puertos distintos. Modifique el modelo propuesto en el punto anterior para incluir esta nueva condición.

### Ejercicio 18

Una empresa petrolera tiene dos refinerías. La refinería 1 tiene un costo diario de operación de USD 25M y produce, diariamente, 4 tanques de petróleo de alto octanaje, 3 tanques de medio octanaje y 2 tanques de bajo octanaje. La refínería 2 es más moderna: Cuesta USD 20M al día para operar y produce 3 tanques de alto octanaje, 4 tanques de medio octanaje y 5 tanques de bajo octanaje, cada día.

La compañia tiene ordenes por 250 tanques de alto octanaje, 270 de mediano octanaje y 300, de bajo octanaje. Para minimizar el costo total de la operación se ha planteado el siguiente modelo de optimización, donde $x_1$ y $x_2$ representan el número de días que operan las refinerías 1 y 2, respectivamente:

$$min \quad 25x_1 + 20x_2$$

s.a.

$$4x_1 + 3x_2 \ge 250$$
$$3x_1 + 4x_2 \ge 270$$
$$2x_1 + 5x_2 \ge 300$$
$$x_1; x_2 \ge 0$$

a. (5 puntos) Escriba la forma estándar del problema.

b. (10 puntos) Considere la solución básica factible que corresponde a operar sólo la refinería 1 durante 150 días (y ninguno, la refinería 2). Utilizando el criterio de optimalidad, muestre que esta no es una solución óptima del problema.

c. (15 puntos) Encuentre el óptimo utilizando el método simplex. Tome como punto inicial la solución básica provista en el item anterior.